api/src/main/java/com/spotify/ffwd/util/HighFrequencyDetector.java

```diff
@@ -0,0 +1,157 @@
+/*-
+ * -\-\-
+ * FastForward API
+ * --
+ * Copyright (C) 2020 Spotify AB
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+
+package com.spotify.ffwd.util;
+
+import com.google.inject.Inject;
+import com.google.inject.name.Named;
+import com.spotify.ffwd.model.Metric;
+import com.spotify.ffwd.statistics.HighFrequencyDetectorStatistics;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class HighFrequencyDetector {
+
+  /** Allow to drop high frequency metrics. */
+  @Inject
+  @Named("dropHighFrequencyMetric")
+  Boolean dropHighFrequencyMetric;
+
+  /** Minimum number of milliseconds allowed between data points. */
+  @Inject
+  @Named("minFrequencyMillisAllowed")
+  Integer minFrequencyMillisAllowed;
+
+  /** Minimum number of times high frequency detected before metrics are dropped. */
+  @Inject
+  @Named("minNumberOfTriggers")
+  Integer minNumberOfTriggers;
+
+  /** Milliseconds high frequency triggers are refreshed. */
+  @Inject
+  @Named("highFrequencyDataRecycleMS")
+  Long highFrequencyDataRecycleMS;
+
+  @Inject
+  private HighFrequencyDetectorStatistics statistics;
+
+  /** High frequency metrics counter. */
+  // TODO review to use ConcurrentHashMap vs AtomicReference
+  final AtomicReference<Map<Integer, Integer>> highFrequencyMetrics =
+      new AtomicReference<>(new HashMap<>());
+
+  final AtomicLong highFrequencyTriggersTS;
+  private AtomicBoolean highFrequencyRecycleLock;
+
+  /* TODO
+      1. unit tests
+          a. sendBatch with many points
+          b. send both with many points
+          c. check swap of high freq hashmap
+   */
+
+  @Inject
+  public HighFrequencyDetector() {
+    this.highFrequencyTriggersTS = new AtomicLong(System.currentTimeMillis());
+    this.highFrequencyRecycleLock = new AtomicBoolean(false);
+  }
+
+  public List<Metric> detectHighFrequencyMetrics(final List<Metric> metrics) {
+    List<Metric> newList = new ArrayList<>();
+
+    // Groups metrics by metric hash code and finds timestamp deltas of ordered data points.
+    // {hashcode -> -1, hashcode1 -> 10}
+    Map<Integer, Integer> groupedMetrics =
```

Comment:
looks like this Map is only used when `dropHighFrequencyMetric` is true, could you move that into the `if (dropHighFrequencyMetric) { ...` block to avoid computing this when it is not needed?

Reply:
Actually I wanted it to be computed regardless of the flag to drop is true or not. So moved it a bit as we are interested in stats before we start dropping them.